src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -1169,6 +1201,60 @@ public DocumentId AddFile(string fullPath, SourceCodeKind sourceCodeKind, Immuta
                 return documentId;
             }
 
+            public void RenameFile(string originalFilePath, string newFilePath)
+            {
+                if (string.IsNullOrEmpty(originalFilePath))
+                {
+                    throw new ArgumentException($"{nameof(originalFilePath)} isn't a valid path.", nameof(originalFilePath));
+                }
+
+                if (string.IsNullOrEmpty(newFilePath))
+                {
+                    throw new ArgumentException($"{nameof(newFilePath)} isn't a valid path.", nameof(newFilePath));
+                }
+
+                lock (_project._gate)
+                {
+                    if (!_documentPathsToDocumentIds.TryGetValue(originalFilePath, out var documentId))
+                    {
+                        throw new ArgumentException($"'{originalFilePath}' is not a source file of this project.");
+                    }
+
+                    _documentPathsToDocumentIds.Remove(originalFilePath);
+                    _documentPathsToDocumentIds.Add(newFilePath, documentId);
+
+                    documentId = documentId.WithDebugName(newFilePath);
+
+                    // Document is already in the workspace and needs to be renamed
+                    if (_documentAlreadyInWorkspace(_project._workspace.CurrentSolution, documentId))
+                    {
+                        if (_project._activeBatchScopes > 0)
+                        {
+                            _documentsRenamedInbatch.Add((documentId, originalFilePath, newFilePath));
```

Comment:
Do we have to remove entries from this if the document is also removed as a part of a batch? (i.e. somebody doing start batch, add, end batch, start match, rename, remove-with-new-name, end batch)

Reply:
I don't think so. Only because if you do multiple renames then a remove we'll need to handle that correctly 

Rename a -> b
Rename b -> c 
Remove c 

It would take multiple passes to get that correct without somehow making this a directed graph/tree.